### PR TITLE
Add WASI bindings

### PR DESCRIPTION
**Pull request by dcodeIO**
This PR adds bindings for [WASI](https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-overview.md), the WebAssembly System Interface. A test makes sure that it parses and initializes properly, and that struct layouts are matching.

* Documentation reference: https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-api.md
* Implementation references:
  * https://github.com/CraneStation/wasmtime/blob/master/wasmtime-wasi/src/wasm32.rs
  * https://github.com/CraneStation/wasi-sysroot/blob/wasi/libc-bottom-half/headers/public/wasi/core.h

Open questions:
* <del>Will the structs "just work"?</del>
* <del>Some constants use prefixes that seem to be unnecessary because implied by their enclosing enum,, see (resolved) comments below</del>
* <del>Does it have to be `wasi_unstable`?</del>

### CODE REVIEW

File: std/assembly/bindings/wasi.ts [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+  POLL = 28,
+  PWR = 29,
+  /** Bad system call. */
+  SYS = 30
+}
+export type signal_t = u8;
+
+/** Flags determining how to interpret the timestamp provided in `subscription_t::u.clock.timeout. */
+export const enum subclockflags {
+  SUBSCRIPTION_CLOCK_ABSTIME = 1
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi.ts]: `SUBSCRIPTION_CLOCK_ABSTIM` or `CLOCK_ABSTIM`?

File: std/assembly/bindings/wasi.ts [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+  DATA_TRUNCATED = 1
+}
+export type roflags_t = u16;
+
+/** Which channels on a socket to shut down. */
+export const enum sdflags {
+  /** Disables further receive operations. */
+  SHUT_RD = 1,
+  /** Disables further send operations. */
+  SHUT_WR = 2
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi.ts]: `SHUT_RD` or `RD`?
`SHUT_WR` or `WR`?

File: std/assembly/bindings/wasi.ts [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+  TRUNC = 8
+}
+export type oflags_t = u16;
+
+/** Flags provided to `sock_recv`. */
+export const enum riflags {
+  /** Returns the message without removing it from the socket's receive queue. */
+  RECV_PEEK = 1,
+  /** On byte-stream sockets, block until the full amount of data can be returned. */
+  RECV_WAITALL = 2
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi.ts]: `RECV_PEEK` or `PEEK`?
`RECV_WAITALL` or `WAITALL`?

File: std/assembly/bindings/wasi.ts [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+  RSYNC = 8,
+  /** Write according to synchronized I/O file integrity completion. */
+  SYNC = 16
+}
+export type fdflags_t = u16;
+
+/** File descriptor attributes. */
+@unmanaged export class fdstat {
+  /** File type. */
+  fs_filetype: filetype_t;
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi.ts]: Remove `fs_ ` prefix?

File: std/assembly/bindings/wasi.ts [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+/** Identifier for a device containing a file system. Can be used in combination with `inode` to uniquely identify a file or directory in the filesystem. */
+export type device_t = u64;
+
+/** A reference to the offset of a directory entry. */
+export type dircookie_t = u64;
+
+/** A directory entry. */
+@unmanaged export class dirent {
+  /** The offset of the next directory entry stored in this directory. */
+  d_next: dircookie_t;
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi.ts]: Remove `_d` prefix?

File: std/assembly/bindings/wasi.ts [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+/** Relative offset within a file. */
+export type filedelta_t = i64;
+
+/** Non-negative file size or length of a region within a file. */
+export type filesize_t = u64;
+
+/** File attributes. */
+@unmanaged export class filestat {
+  /** Device ID of device containing the file. */
+  st_dev: device_t;
```
Comment (dcodeIO) [inline std/assembly/bindings/wasi.ts]: Remove `st_` prefix?

**Comment by jedisct1** [comment]
It has to be `wasi_unstable`. This is what wasm embedders will use, at least for now.

**Comment by dcodeIO** [comment]
I _think_ this should be as complete as it gets, unless there are further comments that is :)

**Pull request merged**